tmt/steps/prepare/ansible.py

```diff
@@ -166,9 +166,16 @@ def go(
 
                 logger.info('playbook-path', playbook_path, 'green')
 
+            # Check if fmf root exists
+            try:
+                playbook_root = self.step.plan.fmf_root
+            except TypeError:
+                raise PrepareError(
+                    "No fmf root found. Directory is not initialized with `tmt init`")
```

Comment:
Seems needlessly complicated :)

```python
if not self.step.plan.fmf_root:
    raise PrepareError("No fmf root found. Directory is not initialized with `tmt init`")
```

Reply:
I already tried to do it like that but its not working, because fmf_root is not value but a getter, its failing before it can go to inside the if statement.

Reply:
I see. But that's very interesting information: it fails because the getter, declared as `-> Path`, runs `return Path(self.node.root)`, and we realized an fmf node may not always be present. So what you do is scratching the surface, your change will turn "give me fmf root" into an exception in one place only, while everything else trying to access fmf root will be left unsanitized. We should go deeper :)

So, it feels like this check & exception should be raised by `fmf_root` itself, to prevent anyone from accessing `fmf_root` without a punishment. I'm not sure how exactly the exception raised *now* by getter looks like, could you share it? Or feel free to design the test right away.

BTW there's enough blame for more pieces of code, e.g. all these classes, `Core`, `Test`, `Plan`, `Story` - they take fmf node as an input parameter: how can they take an fmf node when we agreed that without an fmf node, there is no fmf tree and therefore no fmf nodes representing plans and tests?? How can we ever get to run a playbook, as directed by a `prepare` step of a plan when there is no fmf root and therefore no plans to recognize?? So there is more to review, basically, all constructs based on fmf are now suspicious when they can get constructed without an fmf root and get as far as to run a playbook :) I would start by checking whether we could put the check & exception into `fmf_root` for now, but eventually, we might check how can we even have `Plan` instance running playbooks when we have no fmf root to start with, and push this check to trigger way sooner.

Reply:
Is the new change enough for now? I tested it and it returns me the exception message
`~/.local/bin/tmt run -vvv prepare --how ansible --playbook packages.yml provision --how virtual` (in directory without tmt init)

Reply:
> , we might check how can we even have Plan instance running playbooks when we have no fmf root to start with, and push this check to trigger way sooner.

But we (technically) do allow to specify the plan via CLI, right? IMO in that case fmf root doesn't need to exist.

Reply:
I'm not sure I follow, how do we specify plan via CLI? `plan --name ...` comes to mind, but that plan must exist first in fmf, and that's all I have.

Reply:
I think, a little test would help a lot. With reproducing, with general understanding. I guess the `extend the test coverage` item shouldn't have been removed from the checklist :grin:

So, I believe, @lukaszachy is right: We can still do quick action from the command line with the default plan. The example provided by @falconizmi reproduces. My quick check confirmed:

```
cd
tmt run provision -h container prepare -h ansible --playbook something.yaml login finish
...
fail: argument should be a str or an os.PathLike object where __fspath__ returns a str, not 'NoneType'
```

So yes, it's true that we are able to run outside of an `fmf` root because when no `fmf` metadata is detected the default plan is used instead.

https://github.com/teemtee/tmt/blob/fb19fb0863104fa0d79499fc57929cbf675c5fed/tmt/base.py#L3361-L3363

Reply:
Hm, the `--playbook` options states:

```
Path or URL of an ansible playbook to run.
The playbook path must be relative to the metadata tree root.
```

We need to decide what to do when the default plan is used. I can see a use case for quickly setting up a guest for experimenting when no `fmf` root is around. So... what about something like this:

```
When no `fmf` root is around, the path should be relative to the current working directory.
```

Reply:
Then we also need to change the type of `self.fmf_root` to be `Optional`, and fix the fallout. The assumption that fmf root always exists is widely spread. Maybe we should have something like `self.anchor_path` that would provide this particular path (`self.fmf_root or Path.cwd()`). Many places may need to perform this “is it there or not?” dance.

Reply:
> Maybe we should have something like `self.anchor_path` that would provide this particular path

Yeah, that sounds good.

Reply:
If I understand correctly, instead of having property `fmt_root`  and its getter in base.py, we will have `anchor_path` property and its getter (that would try to return `fmf_root` and if it fails, it would return `Path.cwd()` ) and I would change only those which used `fmf_root` getter from `base.py` and the rest of `fmf_root` usage we would keep them as they are?

Reply:
Yes, that's what I would do. If objects like `Plan` can exist without an fmf root, then it shouldn't be a sin to access `Plan.fmf_root`, instead an "undefined" value like `None` should be returned.